### PR TITLE
feat: add error types for better validation

### PR DESCRIPTION
**Pull request by plusiv**
## Description:
This PR improves error handling by adding specific error types and ensuring compatibility with `errors.Is` for better validation. The changes include:

- Introduced custom error types:
  - `ErrInvalidUUIDFormat`
  - `ErrInvalidURNPrefix`
  - `ErrInvalidBracketedFormat`
  - `ErrInvalidLength`
- Refactored error handling to use the new error types.
- Updated the `Parse`, `ParseBytes`, and `Validate` functions to wrap errors with additional context using `fmt.Errorf` and `%w`.
- Ensured backward compatibility by maintaining the `IsInvalidLengthError` function.

### CODE REVIEW

File: uuid.go [hunk 1]
```diff
@@ -47,18 +47,16 @@ var (
...
-
-type invalidLengthError struct{ len int }
 
-func (err invalidLengthError) Error() string {
-	return fmt.Sprintf("invalid UUID length: %d", err.len)
-}
+	ErrInvalidUUIDFormat      = errors.New("invalid UUID format")
+	ErrInvalidURNPrefix       = errors.New("invalid urn prefix")
+	ErrInvalidBracketedFormat = errors.New("invalid bracketed UUID format")
+	ErrInvalidLength          = errors.New("invalid UUID length")
```
Comment (bormanp) [inline uuid.go:54]: This is a breaking change as it no longer include the invalid length.  There is no reason to change the underlying type of this error.  This should be

```
type LengthError struct { Length int }
```

File: uuid.go [hunk 2]
```diff
@@ -192,10 +190,12 @@ func Must(uuid UUID, err error) UUID {
 }
 
 // Validate returns an error if s is not a properly formatted UUID in one of the following formats:
-//   xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
-//   urn:uuid:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
-//   xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
-//   {xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}
+//
```
Comment (bormanp) [inline uuid.go:193]: Please do not add formatting changes to a PR that is making actual changes.  In general a PR should either be a cosmetic PR or a functional PR but not both.

File: uuid.go [hunk 3]
```diff
@@ -47,18 +47,16 @@ var (
...
 	pool        [randPoolSize]byte // protected with poolMu
-)
-
-type invalidLengthError struct{ len int }
 
-func (err invalidLengthError) Error() string {
-	return fmt.Sprintf("invalid UUID length: %d", err.len)
-}
+	ErrInvalidUUIDFormat      = errors.New("invalid UUID format")
+	ErrInvalidURNPrefix       = errors.New("invalid urn prefix")
```
Comment (bormanp) [inline uuid.go:52]: This should be a new type:
```
type URNPrefixError struct{ Prefix string }
```

File: uuid_test.go [hunk 4]
```diff
@@ -569,7 +569,7 @@ func TestWrongLength(t *testing.T) {
 func TestIsWrongLength(t *testing.T) {
 	_, err := Parse("12345")
 	if !IsInvalidLengthError(err) {
-		t.Errorf("expected error type is invalidLengthError")
+		t.Errorf("expected error type is ErrInvalidLength")
```
Comment (bormanp) [inline uuid_test.go:572]: Technically this is incorrect at `ErrInvalidLength` is an error and not a type.  This should be something like `"IsInvalidLength returned incorrect type %T", err`

File: uuid_test.go [hunk 5]
```diff
@@ -634,8 +634,10 @@ func TestValidate(t *testing.T) {
 	}
 }
 
-var asString = "f47ac10b-58cc-0372-8567-0e02b2c3d479"
-var asBytes = []byte(asString)
+var (
```
Comment (bormanp) [inline uuid_test.go:637]: This is unrelated to the PR

File: uuid_test.go [hunk 6]
```diff
@@ -900,7 +902,7 @@ func TestVersion7Monotonicity(t *testing.T) {
 type fakeRand struct{}
 
 func (g fakeRand) Read(bs []byte) (int, error) {
-	for i, _ := range bs {
+	for i := range bs {
```
Comment (bormanp) [inline uuid_test.go:905]: This is unrelated to the PR.

File: uuid.go [hunk 7]
```diff
@@ -362,7 +362,7 @@ type UUIDs []UUID
 
 // Strings returns a string slice containing the string form of each UUID in uuids.
 func (uuids UUIDs) Strings() []string {
-	var uuidStrs = make([]string, len(uuids))
+	uuidStrs := make([]string, len(uuids))
```
Comment (bormanp) [inline uuid.go:365]: This is unrelated to the PR.

File: json_test.go [hunk 8]
```diff
@@ -59,12 +59,12 @@ func TestJSONUnmarshal(t *testing.T) {
...
 		"empty": {
 			data:           []byte(`{"ID1": ""}`),
-			expectedError:  invalidLengthError{len: 0},
+			expectedError:  ErrInvalidLength,
 			expectedResult: Nil,
 		},
 		"omitempty": {
 			data:           []byte(`{"ID2": ""}`),
-			expectedError:  invalidLengthError{len: 0},
+			expectedError:  ErrInvalidLength,
```
Comment (bormanp) [inline json_test.go:67]: This is concerning.  It implies that `json.Unmarshal` will now return a different error than it did before.  Previously it returned an error that contained the length of the source.  Are you sure this works?  The code is returning something the results of `fmt.Errorf` which is not `ErrInvalidLength`

**Review by bormanp — CHANGES_REQUESTED** [review]
I understand what you are trying to do and overall it seems like a worthy idea.

Given the number of packages that include this package I am rather paranoid about any functional change.   I don't mind the custom error becoming public as that is an addition and changes nothing.

**Comment by plusiv** [comment]
@bormanp I really appreciate your review and valid suggestions. I've removed some changes that were added unintentionally by my code formatter in my code editor. Additionally I've sightly modified the approach with `ErrInvalidLength` and `ErrInvalidURNPrefix` error types to keep more backward compatibility.

File: uuid.go (hunk 2 above)
Comment (plusiv) [reply uuid.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

File: uuid.go (hunk 3 above)
Comment (plusiv) [reply uuid.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

File: uuid_test.go (hunk 4 above)
Comment (plusiv) [reply uuid_test.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

File: uuid_test.go (hunk 5 above)
Comment (plusiv) [reply uuid_test.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

File: uuid_test.go (hunk 6 above)
Comment (plusiv) [reply uuid_test.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

File: uuid.go (hunk 7 above)
Comment (plusiv) [reply uuid.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

File: json_test.go (hunk 8 above)
Comment (plusiv) [reply json_test.go]: Solved in [54f9572](https://github.com/google/uuid/pull/166/commits/54f95728c20a700436b4f3a8e699a99235748175)

**Review by bormanp — APPROVED** [review]
